x-pack/plugins/lens/public/indexpattern_datasource/dimension_panel/operation_support.ts

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { DatasourceDimensionDropProps } from '../../types';
+import { OperationType } from '../indexpattern';
+import { getAvailableOperationsByMetadata } from '../operations';
+import { IndexPatternPrivateState } from '../types';
+
+export interface OperationSupportMatrix {
+  operationByField: Partial<Record<string, OperationType[]>>;
+  operationWithoutField: OperationType[];
+  fieldByOperation: Partial<Record<OperationType, string[]>>;
+}
+
+type Props = Pick<
+  DatasourceDimensionDropProps<IndexPatternPrivateState>,
+  'layerId' | 'columnId' | 'state' | 'filterOperations'
+>;
+
+// TODO: This code has historically been memoized, as a potentially performance
+// sensitive task. If we can add memoization without breaking the behavior, we should.
+export const getOperationSupportMatrix = (props: Props): OperationSupportMatrix => {
+  const layerId = props.layerId;
+  const currentIndexPattern = props.state.indexPatterns[props.state.layers[layerId].indexPatternId];
+
+  const filteredOperationsByMetadata = getAvailableOperationsByMetadata(
+    currentIndexPattern
+  ).filter((operation) => props.filterOperations(operation.operationMetaData));
+
+  const supportedOperationsByField: Partial<Record<string, OperationType[]>> = {};
+  const supportedOperationsWithoutField: OperationType[] = [];
+  const supportedFieldsByOperation: Partial<Record<OperationType, string[]>> = {};
+
+  filteredOperationsByMetadata.forEach(({ operations }) => {
+    operations.forEach((operation) => {
+      if (operation.type === 'field') {
+        if (supportedOperationsByField[operation.field]) {
+          supportedOperationsByField[operation.field]!.push(operation.operationType);
+        } else {
+          supportedOperationsByField[operation.field] = [operation.operationType];
+        }
+
+        if (supportedFieldsByOperation[operation.operationType]) {
+          supportedFieldsByOperation[operation.operationType]!.push(operation.field);
+        } else {
+          supportedFieldsByOperation[operation.operationType] = [operation.field];
+        }
```

Comment:
nitpick: I always find a bit harder to read these conditionals that handles add or create in different ways. What about check the existence of the entry and just push?

```js
supportedOperationsByField[operation.field] = supportedOperationsByField[operation.field] || [];
supportedOperationsByField[operation.field].push(operation.operationType);

supportedFieldsByOperation[operation.operationType] = supportedFieldsByOperation[operation.operationType] || [];
supportedFieldsByOperation[operation.operationType].push(operation.field);
```

Reply:
I think I understand your point, but I'm not sure your proposal is more clear. I'm going to try using the spread operator, but maybe this is just as bad?

```
        supportedFieldsByOperation[operation.operationType] = [
          ...(supportedFieldsByOperation[operation.operationType] ?? []),
          operation.field,
        ];
```